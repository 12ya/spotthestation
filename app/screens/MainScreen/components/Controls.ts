import { Camera, EventDispatcher, Spherical, TOUCH, Vector2, Vector3, BaseEvent } from "three"
import { NativeTouchEvent } from "react-native"

const STATE = {
  NONE: -1,
  TOUCH_ROTATE: 3,
}

export class Controls extends EventDispatcher {
  object: Camera & {
    fov: number
    top: number
    right: number
    left: number
    bottom: number
    zoom: number
    updateProjectionMatrix: () => void
    isOrthographicCamera?: boolean
    isPerspectiveCamera?: boolean
  }

  enabled: boolean
  target: Vector3
  minDistance: number
  maxDistance: number
  minPolarAngle: number
  maxPolarAngle: number
  minAzimuthAngle: number
  maxAzimuthAngle: number
  enableRotate: boolean
  rotateSpeed: number
  touches: { ONE: TOUCH }
  target0: Vector3
  position0: Vector3
  update: () => boolean
  private changeEvent: BaseEvent
  private startEvent: BaseEvent
  private endEvent: BaseEvent
  private state
  private spherical: Spherical
  private sphericalDelta: Spherical
  private scale
  private panOffset: Vector3
  private rotateStart: Vector2
  private rotateEnd: Vector2
  private rotateDelta: Vector2
  saveState: () => void
  reset: () => void
  private rotateLeft: (angle: number) => void
  private rotateUp: (angle: number) => void
  width: number
  getElementWidth: () => number
  height: number
  getElementHeight: () => number
  private handleTouchStartRotate: (event: NativeTouchEvent) => void
  private handleTouchMoveRotate: (event: NativeTouchEvent) => void
  onTouchStart: (event: NativeTouchEvent) => void
  onTouchMove: (event: NativeTouchEvent) => void
  onTouchEnd: (event: NativeTouchEvent) => void

  constructor(object) {
    super()
    this.object = object
    this.enabled = true
    this.target = new Vector3()
    this.minDistance = 0
    this.maxDistance = Infinity
    this.minPolarAngle = 0 // radians
    this.maxPolarAngle = Math.PI // radians
    this.minAzimuthAngle = -Infinity // radians
    this.maxAzimuthAngle = Infinity // radians
    this.enableRotate = true
    this.rotateSpeed = 0.5
    this.touches = { ONE: TOUCH.ROTATE }
    this.changeEvent = { type: "change" }
    this.startEvent = { type: "start" }
    this.endEvent = { type: "end" }
    this.state = STATE.NONE
    this.spherical = new Spherical()
    this.sphericalDelta = new Spherical()
    this.scale = 1
    this.panOffset = new Vector3()
    this.rotateStart = new Vector2()
    this.rotateEnd = new Vector2()
    this.rotateDelta = new Vector2()

    this.saveState = () => {
      this.target0.copy(this.target)
      this.position0.copy(this.object.position)
    }

    this.reset = () => {
      this.target.copy(this.target0)
      this.object.position.copy(this.position0)
      this.object.updateProjectionMatrix()
      this.dispatchEvent(this.changeEvent)
      this.update()
      this.state = STATE.NONE
    }

    this.rotateLeft = (angle) => {
      this.sphericalDelta.theta -= angle
    }

    this.rotateUp = (angle) => {
      this.sphericalDelta.phi -= angle
    }

    this.width = 0
    this.getElementWidth = () => {
      return this.width
    }
    this.height = 0
    this.getElementHeight = () => {
      return this.height
    }

    this.handleTouchStartRotate = ({ touches }) => {
      if (touches.length === 1) {
        this.rotateStart.set(touches[0].pageX, touches[0].pageY)
      } else {
        const x = 0.5 * (touches[0].pageX + touches[1].pageX)
        const y = 0.5 * (touches[0].pageY + touches[1].pageY)
        this.rotateStart.set(x, y)
      }
    }

    this.handleTouchMoveRotate = ({ touches }) => {
      if (touches.length === 1) {
        this.rotateEnd.set(touches[0].pageX, touches[0].pageY)
      } else {
        const x = 0.5 * (touches[0].pageX + touches[1].pageX)
        const y = 0.5 * (touches[0].pageY + touches[1].pageY)
        this.rotateEnd.set(x, y)
      }
      this.rotateDelta.subVectors(this.rotateEnd, this.rotateStart).multiplyScalar(this.rotateSpeed)
      this.rotateLeft((2 * Math.PI * this.rotateDelta.x) / this.getElementHeight())
      this.rotateUp((2 * Math.PI * this.rotateDelta.y) / this.getElementHeight())
      this.rotateStart.copy(this.rotateEnd)
    }

    this.onTouchStart = (event) => {
      if (this.enabled === false) return

      switch (event.touches.length) {
        case 1:
          switch (this.touches.ONE) {
            case TOUCH.ROTATE:
              if (this.enableRotate === false) return
              this.handleTouchStartRotate(event)
              this.state = STATE.TOUCH_ROTATE
              break
            default:
              this.state = STATE.NONE
          }
          break
        default:
          this.state = STATE.NONE
      }
      if (this.state !== STATE.NONE) {
        this.dispatchEvent(this.startEvent)
      }
    }

    this.onTouchMove = (event) => {
      if (this.enabled === false) return

      switch (this.state) {
        case STATE.TOUCH_ROTATE:
          if (this.enableRotate === false) return
          this.handleTouchMoveRotate(event)
          this.update()
          break
        default:
          this.state = STATE.NONE
      }
    }

    this.onTouchEnd = () => {
      if (this.enabled === false) return
      this.dispatchEvent(this.endEvent)
      this.state = STATE.NONE
    }

    this.target0 = this.target.clone()
    this.position0 = this.object.position.clone()

    this.update = (() => {
      const offset = new Vector3()
      return () => {
        const position = this.object.position
        offset.copy(position).sub(this.target)

        this.spherical.setFromVector3(offset)

        this.spherical.theta += this.sphericalDelta.theta
        this.spherical.phi += this.sphericalDelta.phi

        this.spherical.theta = Math.max(
          this.minAzimuthAngle,
          Math.min(this.maxAzimuthAngle, this.spherical.theta),
        )
        this.spherical.phi = Math.max(
          this.minPolarAngle,
          Math.min(this.maxPolarAngle, this.spherical.phi),
        )
        this.spherical.makeSafe()
        this.spherical.radius *= this.scale
        this.spherical.radius = Math.max(
          this.minDistance,
          Math.min(this.maxDistance, this.spherical.radius),
        )
        this.target.add(this.panOffset)

        offset.setFromSpherical(this.spherical)

        position.copy(this.target).add(offset)
        this.object.lookAt(this.target)
        this.sphericalDelta.set(0, 0, 0)
        this.panOffset.set(0, 0, 0)
        this.scale = 1

        return false
      }
    })()

    this.update()
  }
}
