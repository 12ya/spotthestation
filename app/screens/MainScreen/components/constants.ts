export const GLOBE_RADIUS = 300
export const GLOBE_SEGMENTS = 300
