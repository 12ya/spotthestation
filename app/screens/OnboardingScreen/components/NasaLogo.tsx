import React from "react"
import { Image, ImageStyle } from "react-native"
import { scale } from "../../../theme"

const nasaLogo = require("../../../../assets/images/nasa-logo.png")

export function NasaLogo() {
  return <Image source={nasaLogo} style={$logo} />
}

const $logo: ImageStyle = {
  width: scale(110),
  height: scale(110),
  marginTop: scale(36),
  marginLeft: scale(36),
}
