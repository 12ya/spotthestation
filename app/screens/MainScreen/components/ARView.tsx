import React, { useCallback, useEffect, useState, forwardRef, useMemo } from "react"
import { Image, ImageStyle, View, ViewStyle } from "react-native"
import { ViroARSceneNavigator } from "@viro-community/react-viro"

import { colors } from "../../../theme"
import { Compass } from "./Compass"
import { DirectionCircle } from "./DirectionCircle"
import { RecordingIndicator } from "./RecordingIndicator"
import { ISSSceneAR, emitter } from "./ISSSceneAR"
import { azAltToCartesian, cartesianToAzAlt, normalizeHeading } from "../../../utils/geometry"
import { OrbitPoint } from "../../../services/api"
import { CatmullRomCurve3, Vector3 } from "three"
import { StyleFn, useStyles } from "../../../utils/useStyles"

interface ARViewProps {
  isFullScreen: boolean
  isPathVisible: boolean
  isRecording: boolean
  recordedSeconds: number
  issPath: OrbitPoint[]
  setIsSpotted: (value: boolean) => void
  image?: string
  onImageLoaded?: () => void
}

export const ARView = forwardRef<ViroARSceneNavigator, ARViewProps>(function ARView(
  { isFullScreen, isPathVisible, isRecording, recordedSeconds, issPath, setIsSpotted, image, onImageLoaded },
  ref,
) {
  const { $container, $hudContainer, $image } = useStyles(styles)

  const [issPosition, setIssPosition] = useState<[number, number]>(null)
  const [curve, setCurve] = useState<CatmullRomCurve3>()
  const [curveStartsAt, setCurveStartsAt] = useState(0)
  const [curveEndsAt, setCurveEndsAt] = useState(0)
  const [position, setPosition] = useState([0, 0])
  const onScreenPositionChange = useCallback((pos: [number, number]) => {
    setPosition(pos)
  }, [])

  const viroAppProps = useMemo(
    () => ({
      onScreenPositionChange,
    }),
    [onScreenPositionChange],
  )

  const intialScene = useMemo(
    () => ({
      scene: ISSSceneAR as any,
    }),
    [onScreenPositionChange],
  )

  useEffect(() => {
    emitter.emit("settings", { isPathVisible })
  }, [isPathVisible])

  useEffect(() => {
    if (issPath.length === 0) {
      return
    }

    setCurve(
      new CatmullRomCurve3(
        issPath.map(
          (p) => new Vector3(...azAltToCartesian(normalizeHeading(p.azimuth), p.elevation, 10)),
        ),
      ),
    )

    setCurveStartsAt(new Date(issPath[0].date).valueOf())
    setCurveEndsAt(new Date(issPath[issPath.length - 1].date).valueOf())
  }, [issPath])

  useEffect(() => {
    emitter.emit("issData", { curve, curveStartsAt, curveEndsAt })
  }, [curve])

  useEffect(() => {
    if (!curve) {
      setIssPosition(null)
      return undefined
    }
    const update = () => {
      const t =
        (Date.now() - new Date(issPath[0].date).valueOf()) /
        (new Date(issPath[issPath.length - 1].date).valueOf() - new Date(issPath[0].date).valueOf())

      let point: Vector3
      try {
        point = curve.getPoint(t)
      } catch (e) {
        console.error(e)
        return
      }

      setIssPosition(cartesianToAzAlt([point.x, point.y, point.z]))
    }
    update()

    const timeout = setInterval(update, 10000)
    return () => {
      clearInterval(timeout)
    }
  }, [curve])

  return (
    <View style={$container}>
      <ViroARSceneNavigator
        ref={ref}
        worldAlignment="GravityAndHeading"
        autofocus={true}
        initialScene={intialScene}
        viroAppProps={viroAppProps}
        style={$container}
      />

      { Boolean(image) && (
        <Image
          source={{ uri: image }}
          style={$image as ImageStyle}
          onLoad={onImageLoaded}
        />)}

      {isFullScreen && (
        <DirectionCircle screenX={position[0]} screenY={position[1]} setIsSpotted={setIsSpotted} />
      )}

      <View style={$hudContainer}>
        {Boolean(issPosition) && (
          <Compass issPosition={normalizeHeading(issPosition[0])} isFullScreen={isFullScreen} />
        )}
        {isRecording && <RecordingIndicator recordedSeconds={recordedSeconds} />}
      </View>
    </View>
  )
})

const styles: StyleFn = () => {
  const $container: ViewStyle = {
    flex: 1,
    backgroundColor: colors.backgroundDark,
  }

  const $hudContainer: ViewStyle = {
    position: "absolute",
    alignItems: "center",
    width: "100%",
    top: 0,
  }

  const $image: ViewStyle = {
    width: '100%',
    height: '100%',
    position: 'absolute'
  }

  return { $container, $hudContainer, $image }
}
