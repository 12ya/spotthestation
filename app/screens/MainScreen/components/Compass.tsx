/* eslint-disable react-native/no-inline-styles */
import React, { useState, useEffect } from 'react'
import { View, Text, Image, ViewStyle, ImageStyle, TextStyle } from 'react-native'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import CompassHeading from 'react-native-compass-heading'
import { Icon } from '../../../components'
import { colors } from '../../../theme'
const compassLine = require('../../../../assets/icons/compass-line.png')

const LINE_LENGTH = 260
const LETTER_OFFSET = 10

const normalizeHeading = (heading: number) => {
  if (heading < 0) return 360 + heading
  if (heading > 360) return heading - 360
  return heading
}

const isInHeadingRange = (left, right, heading) => {
  if (right < left) return heading > left || heading < right
  return heading >= left && heading <= right
}

const headingOffset = (h1, h2) => {
  if (h2 < h1) return h2 - (h1 - 360)
  return h2 - h1
}

export const Compass = ({ issPosition, isFullScreen }) => {
  const topInset = useSafeAreaInsets().top
  const [heading, setHeading] = useState(0)

  const left = normalizeHeading(heading - 45)
  const right = normalizeHeading(heading + 45)

  const letterPositions = [
    { letter: 'N', offset: 0 },
    { letter: 'E', offset: 90 },
    { letter: 'S', offset: 180 },
    { letter: 'W', offset: 270 },
  ].filter(letter => isInHeadingRange(left, right, letter.offset))

  useEffect(() => {
    CompassHeading.start(1, (result) => {
      setHeading(Number(result.heading))
    }).catch((err) => {
      console.log(err)
    })

    return () => {
      CompassHeading.stop().catch((err) => {
        console.log(err)
      })
    }
  }, [])

  const issVisible = isInHeadingRange(left, right, issPosition)

  return (
    <View style={[$container, { marginTop: isFullScreen ? topInset + 24 : 0 }]}>
      <Image source={compassLine} style={$line} />
      {issVisible && <Icon icon="iss" size={24} containerStyle={[$issIcon, { marginLeft: (headingOffset(left, issPosition) / 90) * LINE_LENGTH }]} />}
      <View style={$lettersContainer}>
        {letterPositions.map(({ letter, offset }) => {
          const letterPosition = (headingOffset(left, offset) / 90) * LINE_LENGTH
          return (
            <View key={letter} style={[$letterContainer, { marginLeft: letterPosition }]}>
              <Text
                key={letter}
                style={$letter}
              >
                {letter}
              </Text>
              <View style={$verticalLine} />
            </View>
          )
        })}
      </View>
    </View>
  )
}

const $container: ViewStyle = {
  height: 20,
  position: 'absolute',
  top: 0,
  alignSelf: 'center',
  width: LINE_LENGTH
}

const $issIcon: ViewStyle = {
  bottom: -10,
  position: 'absolute'
}

const $letter: TextStyle = {
  color: colors.palette.neutral100,
  marginBottom: 5,
  textAlign: 'left'
}

const $letterContainer: ViewStyle = {
  alignItems: 'center',
  marginBottom: 2
}

const $lettersContainer: ViewStyle = {
  bottom: -10,
  flexDirection: 'row',
  height: 'auto',
  left: 0,
  marginTop: -LETTER_OFFSET
}

const $line: ImageStyle = {
  bottom: 0,
  left: 0,
  position: 'absolute',
  width: LINE_LENGTH,
}

const $verticalLine: ViewStyle = {
  borderColor: colors.palette.neutral100,
  borderWidth: 1,
  height: 6,
  width: 1
}
