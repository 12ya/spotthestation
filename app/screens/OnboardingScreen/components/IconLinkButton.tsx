import React from "react"
import {
  Pressable,
  PressableProps,
  StyleProp,
  ViewStyle,
  ImageStyle,
  TextStyle,
} from "react-native"
import { BlurView } from "expo-blur"
import Animated from "react-native-reanimated"
import { Icon, IconTypes, Text } from "../../../components"
import { colors } from "../../../theme/colors"
import { fontSizes, lineHeights, scale } from "../../../theme"

export interface LinkButtonWithArrowProps extends PressableProps {
  /**
   * The name of the icon
   */
  icon?: IconTypes
  /**
   * An optional style override useful for padding & margin.
   */
  buttonStyle?: StyleProp<ViewStyle>
  viewStyle?: StyleProp<ViewStyle>
  /**
   * An optional icon style.
   */
  imageStyle?: StyleProp<ImageStyle>
  /**
   * An optional text instead of icon.
   */
  text?: string
  /**
   * An optional text style.
   */
  textStyle?: StyleProp<TextStyle>
  /**
   * An optional intensity of button blur.
   */
  blurIntensity?: number
  /**
   * An optional icon color.
   */
  iconColor?: string
  /**
   * An optional icon size.
   */
  iconSize?: number

  backgroundColor?: string
}

export function IconLinkButton(props: LinkButtonWithArrowProps) {
  const {
    buttonStyle: $buttonStyle,
    viewStyle: $viewStyleOverride,
    icon,
    iconColor,
    iconSize,
    imageStyle: $imageStyle,
    textStyle: $textStyle,
    text,
    backgroundColor,
    blurIntensity = 0,
    ...rest
  } = props

  return (
    <Pressable accessibilityRole="imagebutton" {...rest}>
      <Animated.View style={[$viewStyle, $buttonStyle]}>
        <BlurView
          intensity={blurIntensity}
          style={[$viewStyle, $viewStyleOverride, backgroundColor && { backgroundColor }]}
        >
          {icon ? (
            <Icon
              icon={icon}
              size={iconSize || 20}
              color={iconColor || colors.palette.neutral100}
              style={$imageStyle}
            />
          ) : (
            <Text text={text} style={[$text, $textStyle]} />
          )}
        </BlurView>
      </Animated.View>
    </Pressable>
  )
}

const $viewStyle: ViewStyle = {
  position: "relative",
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
  width: scale(42),
  height: scale(42),
  backgroundColor: colors.palette.neutral350,
  borderRadius: scale(100),
  overflow: "hidden",
}

const $text: TextStyle = {
  color: colors.palette.neutral100,
  fontSize: fontSizes[28],
  lineHeight: lineHeights[28],
}
