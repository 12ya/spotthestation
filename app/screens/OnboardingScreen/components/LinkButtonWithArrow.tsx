import React from "react"
import {
  Pressable,
  PressableProps,
  PressableStateCallbackType,
  StyleProp,
  ViewStyle,
  ImageStyle
} from "react-native"
import { Icon } from "../../../components"
import { colors } from "../../../theme/colors"

export interface ButtonAccessoryProps {
  style: StyleProp<any>
  pressableState: PressableStateCallbackType
}

export interface LinkButtonWithArrowProps extends PressableProps {
  /**
   * An optional style override useful for padding & margin.
   */
  size?: StyleProp<ViewStyle>
  /**
   * An optional icon style.
   */
  imageStyle?: StyleProp<ImageStyle>
}

export function LinkButtonWithArrow(props: LinkButtonWithArrowProps) {
  const {
    size: $size,
    imageStyle: $imageStyle,
    ...rest
  } = props

  return (
    <Pressable style={[$viewStyle, $size]} accessibilityRole="button" {...rest}>
      <Icon icon="back" size={20} color={colors.palette.neutral100} style={$imageStyle} />
    </Pressable>
  )
}

const $viewStyle: ViewStyle = {
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
  width: 42,
  height: 42,
  backgroundColor: colors.palette.neutral350,
  borderRadius: 100,
}
