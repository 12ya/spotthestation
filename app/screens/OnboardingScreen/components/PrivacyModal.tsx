/* eslint-disable react-native/no-inline-styles */
import React from "react"
import {
  ViewStyle,
  View,
  PressableProps,
  TextStyle,
  Linking,
  Pressable,
  Image,
  ImageStyle,
} from "react-native"
import { Button, Text } from "../../../components"
import { fontSizes, lineHeights, scale, typography } from "../../../theme"
import { colors } from "../../../theme/colors"
import { normalizeHeight } from "../../../utils/normalizeHeight"

const image = require("../../../../assets/images/image.png")

export interface PrivacyModalProps {
  onPressSkip?: PressableProps["onPress"]
  onPressAgree?: PressableProps["onPress"]
}

export function PrivacyModal({ onPressSkip, onPressAgree }: PrivacyModalProps) {
  return (
    <View
      accessible
      accessibilityLabel="coach mark"
      accessibilityHint="coach mark"
      accessibilityRole="text"
      style={[$coachModalBodyContainer, { marginTop: normalizeHeight(0.1) }]}
    >
      <Text tx="privacy.title" style={$privacyTitle} />
      <Text tx="privacy.body" style={$body} />
      <Image source={image} style={$image} resizeMode="cover" />
      <View style={$buttonsContainer}>
        <Button
          accessible
          accessibilityLabel="skip button"
          accessibilityHint="skip coach mark"
          tx="privacy.skip"
          textStyle={$skipButtonText}
          style={$skipButton}
          pressedStyle={$skipButton}
          onPress={onPressSkip}
        />
        <Button
          accessible
          accessibilityLabel="next button"
          accessibilityHint="next coach mark"
          tx="privacy.agree"
          textStyle={$nextButtonText}
          style={$nextButton}
          pressedStyle={$nextButton}
          onPress={onPressAgree}
        />
      </View>
      <Pressable
        onPress={() =>
          Linking.openURL("https://www.nasa.gov/about/highlights/HP_Privacy.html#privacy")
        }
        style={{ marginTop: scale(10), alignSelf: "center" }}
      >
        <Text tx="privacy.policy" style={[$tip, { color: colors.palette.neutral100 }]} />
      </Pressable>
    </View>
  )
}

const $image: ImageStyle = {
  alignSelf: "center",
  width: "100%",
  height: normalizeHeight(0.2),
  marginVertical: scale(24),
}

const $buttonsContainer: ViewStyle = {
  flexDirection: "row",
  justifyContent: "space-between",
  width: "100%",
  alignSelf: "flex-end",
}

const $skipButton: ViewStyle = {
  backgroundColor: "transparent",
  borderWidth: 0,
  height: scale(56),
  minWidth: scale(140),
}

const $skipButtonText: TextStyle = {
  fontFamily: typography.primary.medium,
  fontSize: fontSizes[18],
  lineHeight: lineHeights[22],
  color: colors.palette.neutral100,
}

const $nextButton: ViewStyle = {
  height: scale(56),
  backgroundColor: colors.palette.neutral100,
  borderRadius: scale(28),
  borderWidth: 0,
  minWidth: scale(140),
}

const $nextButtonText: TextStyle = {
  fontFamily: typography.primary.medium,
  fontSize: fontSizes[18],
  lineHeight: lineHeights[22],
  color: colors.palette.buttonBlue,
}

const $body: TextStyle = {
  fontFamily: typography.primary.normal,
  fontSize: fontSizes[20],
  lineHeight: lineHeights[24],
  color: colors.palette.neutral100,
  paddingBottom: 10,
  paddingLeft: 5,
}

const $coachModalBodyContainer: ViewStyle = {
  backgroundColor: colors.palette.buttonBlue,
  borderRadius: scale(16),
  paddingVertical: 36,
  paddingHorizontal: 30,
  width: "100%",
}

const $tip: TextStyle = {
  color: colors.palette.neutral450,
  fontSize: fontSizes[18],
  fontFamily: typography.primary.light,
  lineHeight: lineHeights[22],
  textDecorationLine: "underline",
  marginTop: scale(16),
}

const $privacyTitle: TextStyle = {
  color: colors.palette.neutral250,
  fontSize: fontSizes[36],
  fontFamily: typography.primary.normal,
  lineHeight: lineHeights[44],
  paddingBottom: scale(24),
}
