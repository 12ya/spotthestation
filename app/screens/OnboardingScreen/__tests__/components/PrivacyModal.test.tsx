/* eslint-disable @typescript-eslint/no-unsafe-call */
import React from "react"
import { render, fireEvent } from "@testing-library/react"
import renderer from "react-test-renderer"
import { PrivacyModal } from "../../components/PrivacyModal"
import { Linking } from "react-native"

describe("PrivacyModal", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<PrivacyModal />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("calls onPressSkip when skip button is pressed", () => {
    const onPressSkipMock = jest.fn()
    const { getByText } = render(<PrivacyModal onPressSkip={onPressSkipMock} />)
    const skipButton = getByText("privacy.skip undefined")

    ;(fireEvent as any).click(skipButton)

    expect(onPressSkipMock).toHaveBeenCalledTimes(1)
  })

  it("calls onPressAgree when agree button is pressed", () => {
    const onPressAgreeMock = jest.fn()
    const { getByText } = render(<PrivacyModal onPressAgree={onPressAgreeMock} />)
    const agreeButton = getByText("privacy.agree undefined")

    ;(fireEvent as any).click(agreeButton)

    expect(onPressAgreeMock).toHaveBeenCalledTimes(1)
  })

  it("opens the privacy policy URL when privacy policy text is pressed", () => {
    const { getByText } = render(<PrivacyModal />)
    const privacyPolicyText = getByText("privacy.policy undefined")

    jest.spyOn(Linking, "openURL")
    ;(fireEvent as any).click(privacyPolicyText)

    expect((Linking as any).openURL).toHaveBeenCalledWith("https://www.spotthestation.org")
  })
})
