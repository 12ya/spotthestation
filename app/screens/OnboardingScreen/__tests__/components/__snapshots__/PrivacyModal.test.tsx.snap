// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivacyModal renders correctly 1`] = `
<View
  accessibilityHint="coach mark"
  accessibilityLabel="coach mark"
  accessibilityRole="text"
  accessible={true}
  style={
    [
      {
        "backgroundColor": "#0B3D91",
        "borderRadius": 23,
        "paddingHorizontal": 30,
        "paddingVertical": 36,
        "width": "100%",
      },
      {
        "marginTop": 133.4,
      },
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      [
        {},
        [
          {
            "fontSize": 23,
            "lineHeight": 35,
          },
          {
            "fontFamily": "Helvetica Neue",
          },
          {
            "color": "#191015",
          },
        ],
        undefined,
        undefined,
        {
          "color": "#DEDDDE",
          "fontFamily": "Helvetica Neue",
          "fontSize": 52,
          "lineHeight": 63,
          "paddingBottom": 35,
        },
      ]
    }
  >
    privacy.title undefined
  </Text>
  <Text
    allowFontScaling={false}
    style={
      [
        {},
        [
          {
            "fontSize": 23,
            "lineHeight": 35,
          },
          {
            "fontFamily": "Helvetica Neue",
          },
          {
            "color": "#191015",
          },
        ],
        undefined,
        undefined,
        {
          "color": "#FFFFFF",
          "fontFamily": "Helvetica Neue",
          "fontSize": 29,
          "lineHeight": 35,
          "paddingBottom": 10,
          "paddingLeft": 5,
        },
      ]
    }
  >
    privacy.body undefined
  </Text>
  <RCTImageView
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    internal_analyticTag={null}
    resizeMode="cover"
    source={
      [
        {
          "height": undefined,
          "uri": undefined,
          "width": undefined,
        },
      ]
    }
    style={
      {
        "alignSelf": "center",
        "height": 266.8,
        "marginVertical": 35,
        "overflow": "hidden",
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "alignSelf": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <View
      accessibilityHint="skip coach mark"
      accessibilityLabel="skip button"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          [
            {
              "alignItems": "center",
              "borderRadius": 6,
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 80,
              "overflow": "hidden",
              "paddingHorizontal": 17,
              "paddingVertical": 17,
            },
            {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#B6ACA6",
              "borderWidth": 1,
            },
          ],
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
            "height": 80,
            "minWidth": 201,
          },
          false,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        style={
          [
            {},
            [
              {
                "fontSize": 23,
                "lineHeight": 35,
              },
              {
                "fontFamily": "Helvetica Neue",
              },
              {
                "color": "#191015",
              },
            ],
            undefined,
            undefined,
            [
              {
                "flexGrow": 0,
                "flexShrink": 1,
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 23,
                "lineHeight": 29,
                "textAlign": "center",
                "zIndex": 2,
              },
              {
                "color": "#FFFFFF",
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 26,
                "lineHeight": 32,
              },
              false,
            ],
          ]
        }
      >
        privacy.skip undefined
      </Text>
    </View>
    <View
      accessibilityHint="next coach mark"
      accessibilityLabel="next button"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          [
            {
              "alignItems": "center",
              "borderRadius": 6,
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 80,
              "overflow": "hidden",
              "paddingHorizontal": 17,
              "paddingVertical": 17,
            },
            {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#B6ACA6",
              "borderWidth": 1,
            },
          ],
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 40,
            "borderWidth": 0,
            "height": 80,
            "minWidth": 201,
          },
          false,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        style={
          [
            {},
            [
              {
                "fontSize": 23,
                "lineHeight": 35,
              },
              {
                "fontFamily": "Helvetica Neue",
              },
              {
                "color": "#191015",
              },
            ],
            undefined,
            undefined,
            [
              {
                "flexGrow": 0,
                "flexShrink": 1,
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 23,
                "lineHeight": 29,
                "textAlign": "center",
                "zIndex": 2,
              },
              {
                "color": "#0B3D91",
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 26,
                "lineHeight": 32,
              },
              false,
            ],
          ]
        }
      >
        privacy.agree undefined
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "center",
        "marginTop": 15,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        [
          {},
          [
            {
              "fontSize": 23,
              "lineHeight": 35,
            },
            {
              "fontFamily": "Helvetica Neue",
            },
            {
              "color": "#191015",
            },
          ],
          undefined,
          undefined,
          [
            {
              "color": "#ADADAE",
              "fontFamily": "HelveticaNeue-Light",
              "fontSize": 26,
              "lineHeight": 32,
              "marginTop": 23,
              "textDecorationLine": "underline",
            },
            {
              "color": "#FFFFFF",
            },
          ],
        ]
      }
    >
      privacy.policy undefined
    </Text>
  </View>
</View>
`;
